[DPE-5530] Hardening Userguide (#249)
canonical/kafka-operator: docs/explanation/e-hardening.md

# Security Hardening Guide

> This document provides all the guidance and instructions needed to achieve 

Author: I actually copied this sentence from your document :D

a secure deployment of the Charmed Kafka product, including setting up and managing a secure environment.
The document is divided into the following sections:

1. Environment, outlining the recommendation for deploying a secure environment
2. Applications, outlining the product features that enable a secure deployment of a Kafka cluster
3. Additional resources, providing any further information about security and compliance

## Environment

> The environment where applications operate can be divided in three components:

Reviewer: We need to either add OS back in the list below, or replace `three` with `two`.
I'd prefer to add OS to the Environment, but see my comment below about talking specifically about the Host machine's OS and VM's OS.

1. Cloud
2. Juju 

### Cloud

> Charmed Kafka can be deployed on top of several clouds and virtualization layers. 

Reviewer: ```suggestion
Charmed Kafka can be deployed on top of multiple different cloud platforms:
```

Author: are you sure that we should use tables as bullet points? I would have agreed with you if we had bullet points but since we use a table, it would seem more appropriate to use a sentence. Is there any discussion on a style guide about this?

Reviewer: Are you referring to the usage of an introductory sentence with a colon before the table?
It is the usual way to do it, for example, see our [Style guide](https://docs.ubuntu.com/styleguide/en#spelling).
We are not using a table as bullet points (list) here, but they are similar in some ways.
Consider a list to be a one-dimensional table or a table to be a two-dimensional list =).

> The following table provides references for the security documentation for the 
> main supported cloud platforms.

Reviewer: ```suggestion
```

| AWS       | [Best Practices for Security, Identity and Compliance](https://aws.amazon.com/architecture/security-identity-compliance), [AWS security credentials](https://docs.aws.amazon.com/IAM/latest/UserGuide/security-creds.html#access-keys-and-secret-access-keys) | 
| Azure     | [Azure security best practices and patterns](https://learn.microsoft.com/en-us/azure/security/fundamentals/best-practices-and-patterns), [Managed identities for Azure resource](https://learn.microsoft.com/en-us/entra/identity/managed-identities-azure-resources/)                                                                            |

### Juju 

Juju is the component responsible for orchestrating the entire lifecycle, from deployment to Day 2 operations, of 
all applications. Therefore, it is imperative that it is set up securely. Please refer to the Juju documentation for more information on:

* [Juju security](https://discourse.charmhub.io/t/juju-security/15684)

> * [How to harder your deployment](https://juju.is/docs/juju/harden-your-deployment)

Reviewer: ```suggestion
* [How to harden your deployment](https://juju.is/docs/juju/harden-your-deployment)
```

#### Cloud credentials

When configuring the cloud credentials to be used with Juju, ensure that the users have correct permissions to operate at the required level. 
Juju superusers responsible for bootstrapping and managing controllers require elevated permissions to manage several kind of resources, such as
virtual machines, networks, storages, etc. Please refer to the references below for more information on the policies required to be used depending on the cloud. 

| Cloud     | Cloud user policies                                                                                                                                                                                                                            |
|-----------|------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|

> | OpenStack |                                                                                                                                                                                                                                                |

Reviewer: Can we add something for the OpenStack?

Author: I had already asked a person in openstack for such information, but he hasn't gotten back to me with this yet. As soon as I have this info, I will update. However there is also the chance that that information may not exists. What should we do in that case? Just putting "N/A"?

Reviewer: Yes, `TBA`, `TBD`, or `N/A` is much better than an empty cell.

| AWS       | [Juju AWS Permission](https://discourse.charmhub.io/t/juju-aws-permissions/5307), [AWS Instance Profiles](https://discourse.charmhub.io/t/using-aws-instance-profiles-with-juju-2-9/5185), [Juju on AWS](https://juju.is/docs/juju/amazon-ec2) | 
| Azure     | [Juju Azure Permission](https://juju.is/docs/juju/microsoft-azure), [How to use Juju with Microsoft Azure](https://discourse.charmhub.io/t/how-to-use-juju-with-microsoft-azure/15219)                                                                                                  |

#### Juju users

> It is very important that the different juju users are set up with minimal permission depending on the scope for their operations. 

Reviewer: Note for future plans: We have a task in our backlog to create a page about Users in the Explanation section. It will be nice to add such a page with more specific information and then link to it here.

Author: Totally! I agree!

Please refer to the [User access levels](https://juju.is/docs/juju/user-permissions) documentation for more information on the access level and corresponding abilities 
that the different users can be granted. 

> Juju user credentials must be stored securely and rotated regularly to prevent leakage and unauthorized access. 

Reviewer: ```suggestion
Juju user credentials must be stored securely and rotated regularly to prevent unauthorized access. 
```

Author: why? The point of rotating credentials is related to the fact that credentials may leak

Reviewer: Correct me if I'm wrong but my understanding is the following.
1. "Leak" is an unofficial term. Secrets can be compromised. 
2. Anyone else using your "leaked" credentials is performing unauthorized access.
3. You can't prevent a "leak" by rotating. You can prevent unauthorized access with the leaked credentials (by rotating post-factum either specifically to address compromised credentials or by schedule).

Author: Yes, that's fine. I would then suggest:

> Juju user credentials must be stored securely and rotated regularly to limit chances of unauthorized access due to credentials leakage. 

Unauthorized access can happen for different reasons. I don't think we should claim that rotation prevent all unauthorized access.

Reviewer: Looks much better this way. If you really want to put the word "leakage" there it is probably the best way to do it.
I'll add it as a proper suggestion because that enables me to approve this PR with the suggestion. And if this suggestion is merged it should not invalidate my approval.
```suggestion
Juju user credentials must be stored securely and rotated regularly to limit the chances of unauthorized access due to credentials leakage.
```
Alternatively, I'd consider using "due to compromised credentials".

## Applications

In the following we provide guidance on how to harden your deployment using:

> 1. Operating System

Author: I'm not sure what you mean with the host machine. People could be running everything on the cloud, and not having a host. I believe it is sensible to have the OS in the applications, since these are connected to the Kafka and ZooKeeper deployments, and for instance could be even different between different charms. 

We could have a section on "Machine", but these are tightly connected and spawn by juju, not the environment that needs to be set.

> 1. Operating System

Reviewer: Yes, this topic can be confusing, but it is really important to set the borders of our product here. It's usually a very fine line, as it's about layers of software and responsibility. I assumed our product is the charmed operator/bundle and everything else, including snaps, images, Juju, etc. is the environment. I understand that it is different in Juju, where each application can be deployed on its own OS image. Do we consider the OS image to be a part of our distribution/bundle or merely a dependency?
I recommend discussing this particular issue with our Security team later. For now, I'm happy to proceed without changes.

2. Kafka and ZooKeeper Security Upgrades
3. Encryption 
4. Authentication
5. Monitoring and Auditing

### Operating System

> Charmed Kafka and Charmed ZooKeeper currently run on top of Ubuntu 22.04. Deploy a [Landscape Client Charm]() in order to 

Reviewer: Charms run on VMs that use Ubuntu 22.04. Can we be a little bit more accurate here? The K8s document will differ, obviously. And that difference is important to take present here.

Author: Yes, I'll create a document for K8s. This one is only for machines

> Charmed Kafka and Charmed ZooKeeper currently run on top of Ubuntu 22.04. Deploy a [Landscape Client Charm]() in order to 

Author: Oh yes, I'll add it

connect the underlying VM to a Landscape User Account to manage security upgrades and integrate Ubuntu Pro subscriptions. 

### Kafka and ZooKeeper Security Upgrades

Charmed Kafka and Charmed ZooKeeper operators install a pinned revision of the [Charmed Kafka snap](https://snapcraft.io/charmed-kafka)
and [Charmed ZooKeeper snap](https://snapcraft.io/charmed-zookeeper), respectively, in order to provide reproducible and secure environments. 
New versions of Charmed Kafka and Charmed ZooKeeper may be released to provide patching of vulnerabilities (CVEs). 
It is important to refresh the charm regularly to make sure the workload is as secure as possible. 

> For more information on how to refresh the charm, please refer to the [how-to upgrade](https://charmhub.io/kafka/docs/h-upgrade) documentation.

Reviewer: ```suggestion
For more information on how to refresh the charm, see the [how-to upgrade](https://charmhub.io/kafka/docs/h-upgrade) guide.
```

### Encryption

> Charmed Kafka must be deployed with security enabled. This requires to deploy Kafka and ZooKeeper and relate them to one of the TLS certificate operator charms. 

Reviewer: > Charmed Kafka must be deployed with security enabled.

That sounds like we can enable **security** in our config or something.

```suggestion
Charmed Kafka must be deployed with encryption enabled. 
To do that, you need to relate Kafka and ZooKeeper charms to one of the TLS certificate operator charms.  
```

Please refer to the [Charming Security page](https://charmhub.io/topics/security-with-x-509-certificates) for more information on how to select the right certificate
provider for your use-case. 

> For more information, please refer to the [how to enable encryption](https://charmhub.io/kafka/docs/h-enable-encryption) user guide.

Reviewer: ```suggestion
For more information on encryption setup, see the [How to enable encryption](https://charmhub.io/kafka/docs/h-enable-encryption) guide.
```

### Authentication

Charmed Kafka supports the following authentication layers:

> 1. [SCRAM-based SASL Authentication](/t/charmed-kafka-how-to-manage-app/10285)

Reviewer: ```suggestion
1. [SCRAM-based SASL Authentication](/t/10285)
```
This is the recommended way to do the links to the Discourse. It's shorter and easier to use, but lacks context of the link in the URL, but it makes it more resilient to errors and changes. It is not mandatory, but it is recommended.

Feel free to ignore or apply to all similar links in the file.

Author: Really? I didn't know that was the recommendation. Do you have a link for this?

If there is not a strong guidance, I would honestly prefer to have the slug which provides a bit more information than an ID only.

Reviewer: That's not a requirement, so feel free to use the full version with the slug. 
I don't have any written evidence ready at the moment, I just recall a discussion we had with other TAs on this. The main reason for the shorter version being the recommended one, I believe, was that it looks much simpler (encourages people to contribute) and less prone to errors.

2. [certificate-base Authentication (mTLS)](/t/create-mtls-client-credentials/11079)
3. OAuth Authentication using [Hydra](/t/how-to-connect-to-kafka-using-hydra-as-oidc-provider/14610) or [Google](/t/how-to-connect-to-kafka-using-google-as-oidc-provider/14611)

Each combination of authentication scheme and encryption is associated to the dedicated listener and it maps to a well-defined port. 
Please refer to the [listener reference documentation](/t/charmed-kafka-documentation-reference-listeners/13264) for more information. 

### Monitoring and Auditing

Charmed Kafka provides native integration with the [Canonical Observability Stack (COS)](https://charmhub.io/topics/canonical-observability-stack).

> To reduce the blast radius of infrastructure disruptions, the general recommendation is to deploy COS and the observed application into 

Reviewer: ```suggestion
To reduce the damage of infrastructure disruptions, the general recommendation is to deploy COS and the observed application into 
```

Author: Uhm, I actually took this phrasing from observability docs: https://charmhub.io/topics/canonical-observability-stack/reference/best-practices

separate environments, isolated one another. Refer to the [COS production deployments best practices](https://charmhub.io/topics/canonical-observability-stack/reference/best-practices)
for more information. 

> Refer to [How-to] user guide for more information on:

Reviewer: Is that an empty link?

* [how to integrate the Charmed Kafka deployment with COS](/t/charmed-kafka-how-to-enable-monitoring/10283)
* [how to customise the alerting rules and dashboards](/t/charmed-kafka-documentation-how-to-integrate-custom-alerting-rules-and-dashboards/13431)

> External user access to Kafka is logged to the `kafka-authorizer.log` that is pushes to Loki and exposed via Grafana.

Reviewer: That's the only time we mention Loki. Can we add a link to it? Because if someone needs to google it they will find lots of Marvel stuff instead. Alternatively, we can call it a bit more specific, like "Grafana Loki".

Author: grafana is a different thing, I don't think "Grafana Loki" makes much sense. Anyhow I have rephrased the sentence and provided links
